### PR TITLE
Add a 'requireForBlockBody' modifier to the 'arrow-parens' rule (fixes #6557)

### LINKED ISSUE #6557: [arrow-parens] Add a "when-brace" option
Our team would like to adopt the AirBnb best practice for arrow functions parens but it appears the current options of this rule doesn't fit it. Since we want to rely on automated coding style rules only and found out it would be useful for others too, here is the PR that fulfill it.

See https://github.com/airbnb/javascript#arrows--one-arg-parens
Discussion: https://github.com/SimpliField/eslint-config-simplifield/issues/14

- The version of ESLint you are using

2.11.1
https://github.com/SimpliField/eslint-config-simplifield/blob/master/package.json#L20

- The rule you want to change

'arrow-parens'

- The code you want to be flagged as incorrect
```js
a => { a } // (1)
(a) => a // (2)
```
- What happens when the rule is applied to the code without your change

with 'as-needed' (1) is valid

with 'always' (2) is valid

### PR DESCRIPTION
**Pull request by nfroidure**
The when-brace option allows to require parens for arrow functions whose body is
 wrapped into curly braces like the following:
 ```js
 (a) => {}
 ```

Otherwise, its behavior remains identical then with the as-needed option.

### CODE REVIEW

**Comment by SebastienElet** [comment]
👍

**Comment by kaicataldo** [comment]
Issue is now accepted

**Comment by kaicataldo** [comment]
@nfroidure Thanks for contributing to ESLint! We're ready to start moving on this now that the enhancement has been discussed and accepted. The team ended up deciding on a different configuration option than the one you have here - would you be willing to update the PR so that it reflects the issue? Let me know if you have any questions!

**Comment by nfroidure** [comment]
@kaicataldo thought you were on it, i can do it as well, just let me know to avoid doing it twice.

**Comment by kaicataldo** [comment]
Ah, understand the confusion. As a champion for this enhancement it's up to me to either implement or to help someone else implement it. We'd love your contribution, if you're willing :)

If you don't have the time or are not interested, I can definitely take this on.

**Comment by nfroidure** [comment]
@kaicataldo done ;). I amended the commit and renamed it, just tell me if i also have to change the branch name.

**Comment by nfroidure** [comment]
@eslintbot For sure, wait a minute ⌚

**Comment by kaicataldo** [comment]
No worries on the branch name :) The team will start reviewing and we'll leave comments here, should we feel anything needs to be changed/updated. Haven't had a chance to look at the PR yet but will try to do so ASAP. Thanks for contributing!

File: docs/rules/arrow-parens.md [hunk 1]
```diff
@@ -49,13 +49,16 @@ a => {}
...
 
-The rule takes one option, a string, which could be either `"always"` or `"as-needed"`. The default is `"always"`.
+This rule has a string option:
+
+* `"always"` (default) requires parens in all cases.
+* `"as-needed"` allows ommitting parens in some cases.
+
+ This rule has an object option for exceptions to the `"as-needed"` option:
 
 You can set the option in configuration like this:
```
Comment (btmills) [inline docs/rules/arrow-parens.md]: Looks like we don't need this line anymore.

File: docs/rules/arrow-parens.md [hunk 2]
```diff
@@ -176,4 +179,45 @@ a.then(foo => { if (true) {}; });
...
 ([a, b]) => a;
 ({a, b}) => a;
-```
\ No newline at end of file
+```
+
+
+### requireForBlockBody
+
+When the rule is set to `"as-needed", { "requireForBlockBody": true }` the following patterns are considered problems:
```
Comment (btmills) [inline docs/rules/arrow-parens.md]: To help our documentation generator, can you change this line to read

> Examples of **incorrect** code for the `{ "requireForBlockBody": true }` option:

The line introducing the correct examples below should be similar, saying "correct" instead of "incorrect".

File: docs/rules/arrow-parens.md [hunk 3]
```diff
@@ -176,4 +179,45 @@ a.then(foo => { if (true) {}; });
...
+```js
+/*eslint arrow-parens: [2, "as-needed", { "requireForBlockBody": true }]*/
+/*eslint-env es6*/
+
+(a) => {};
+(a) => {'\n'};
+() => {};
+a => a;
+a.then((foo) => {});
+a.then(foo => { if (true) {}; });
```
Comment (btmills) [inline docs/rules/arrow-parens.md]: Doesn't this example need parens?

File: lib/rules/arrow-parens.js [hunk 4]
```diff
@@ -20,7 +20,16 @@ module.exports = {
 
         schema: [
             {
-                enum: ["always", "as-needed"]
+                enum: ["always", "as-needed", "when-brace"]
```
Comment (btmills) [inline lib/rules/arrow-parens.js]: The `"when-brace"` option isn't needed anymore.

**Comment by btmills** [comment]
Just a handful of minor tweaks, otherwise LGTM. Thanks @nfroidure!

**Comment by ljharb** [comment]
@nfroidure Thank you!! I'm looking forward to enabling this rule in Airbnb's linter config :-)

**Comment by nfroidure** [comment]
@btmills thanks for the review, just updated the PR.

File: docs/rules/arrow-parens.md [hunk 5]
```diff
@@ -176,4 +177,44 @@ a.then(foo => { if (true) {}; });
 (a = 10) => a;
 ([a, b]) => a;
 ({a, b}) => a;
-```
\ No newline at end of file
+```
+
+### requireForBlockBody
+
+Examples of incorrect code for the { "requireForBlockBody": true } option:
```
Comment (btmills) [inline docs/rules/arrow-parens.md]: Oops, it looks like this line and the corresponding "correct" line lost some formatting when you copied them from my previous comment. The words "incorrect" and "correct" need double asterisks, and the code needs backticks:

```md
Examples of **incorrect** code for the `{ "requireForBlockBody": true }` option:
```

Comment (nfroidure) [inline docs/rules/arrow-parens.md]: @btmills Just fixed it. Sorry.

File: docs/rules/arrow-parens.md [hunk 6]
```diff
@@ -49,13 +49,14 @@ a => {}
 
 ## Options
 
-The rule takes one option, a string, which could be either `"always"` or `"as-needed"`. The default is `"always"`.
+This rule has a string option:
```
Comment (kaicataldo) [inline docs/rules/arrow-parens.md]: My preference would be to note at the top level that this also has a second object option and then go into more detail into each option. That way the reader can know at a glance that they should keep going down the page if they're interested in all the options.

File: docs/rules/arrow-parens.md [hunk 7]
```diff
@@ -49,13 +49,14 @@ a => {}
...
 
-You can set the option in configuration like this:
+* `"always"` (default) requires parens in all cases.
+* `"as-needed"` allows ommitting parens in some cases.
```
Comment (kaicataldo) [inline docs/rules/arrow-parens.md]: Can we describe here, even at a high level, what these cases are?

File: docs/rules/arrow-parens.md [hunk 8]
```diff
@@ -176,4 +177,44 @@ a.then(foo => { if (true) {}; });
...
+a.then(foo => {});
+```
+
+Examples of **correct** code for the `{ "requireForBlockBody": true }` option:
+
+```js
+/*eslint arrow-parens: [2, "as-needed", { "requireForBlockBody": true }]*/
+/*eslint-env es6*/
+
+(a) => {};
```
Comment (kaicataldo) [inline docs/rules/arrow-parens.md]: Thoughts on adding `a => ({});` to the correct examples?

File: docs/rules/arrow-parens.md [hunk 9]
```diff
@@ -176,4 +177,44 @@ a.then(foo => { if (true) {}; });
...
+a.then((foo) => { if (true) {}; });
+a((foo) => { if (true) {}; });
+(a, b, c) => a;
+(a = 10) => a;
+([a, b]) => a;
+({a, b}) => a;
+```
+
+The `"as-needed", { "requireForBlockBody": true }` rule is directly inspired by the Airbnb
+ [JS Style Guide](https://github.com/airbnb/javascript#arrows--one-arg-parens).
```
Comment (kaicataldo) [inline docs/rules/arrow-parens.md]: Thoughts on adding this link in a section at the bottom like many of other rules do? Example: https://github.com/eslint/eslint/blob/master/docs/rules/no-debugger.md

File: lib/rules/arrow-parens.js [hunk 10]
```diff
@@ -29,18 +38,63 @@ module.exports = {
...
          */
         function parens(node) {
             var token = sourceCode.getFirstToken(node);
+            var bodyToken = sourceCode.getFirstToken(node.body);
+
+            // "as-needed", { "requireForBlockBody": true }: x => x
+            if (
+                requireForBlockBody &&
+                node.params.length === 1 && node.params[0].type === "Identifier" &&
+                bodyToken.type !== "Punctuator" && bodyToken.value !== "{"
```
Comment (kaicataldo) [inline lib/rules/arrow-parens.js]: I personally think `node.body.type !== 'BlockStatement'` is clearer than `bodyToken.type !== "Punctuator" && bodyToken.value !== "{"`. Thoughts?

File: lib/rules/arrow-parens.js [hunk 11]
```diff
@@ -29,18 +38,63 @@ module.exports = {
...
+                            ], paramToken.value);
+                        }
+                    });
+                }
+                return;
+            }
+
+            if (
+                requireForBlockBody &&
+                bodyToken.type === "Punctuator" && bodyToken.value === "{"
```
Comment (kaicataldo) [inline lib/rules/arrow-parens.js]: Same comment as above - thoughts on checking if `node.body.type` is 'BlockStatement'? Seems like a good idea to rely on info the parser is giving us, since we can.

File: lib/rules/arrow-parens.js [hunk 12]
```diff
@@ -29,18 +38,63 @@ module.exports = {
...
+            ) {
+                if (token.type !== "Punctuator" || token.value !== "(") {
```
Comment (kaicataldo) [inline lib/rules/arrow-parens.js]: Is this just to short-circuit the check if it's not a punctuator? That's cool! I would have written it `token.type !== "Punctuator" || token.value !== "("`, but I like this :)

File: tests/lib/rules/arrow-parens.js [hunk 13]
```diff
@@ -18,30 +18,53 @@ var rule = require("../../../lib/rules/arrow-parens"),
...
     { code: "a => a", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
     { code: "([a, b]) => {}", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
     { code: "({ a, b }) => {}", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
     { code: "(a = 10) => {}", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
     { code: "(...a) => a[0]", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
-    { code: "(a, b) => {}", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } }
+    { code: "(a, b) => {}", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
+
+    // "as-needed", { "requireForBlockBody": true }
+    { code: "() => {}", options: ["as-needed", {requireForBlockBody: true}], parserOptions: { ecmaVersion: 6 } },
```
Comment (kaicataldo) [inline tests/lib/rules/arrow-parens.js]: Can we add tests that ensure `(a) => {};` and `a => ({});` are valid?

File: tests/lib/rules/arrow-parens.js [hunk 14]
```diff
@@ -133,6 +156,44 @@ var invalid = [
...
+        parserOptions: { ecmaVersion: 6 },
+        errors: [{
+            line: 1,
+            column: 1,
+            message: requireForBlockBodyMessage,
+            type: type
+        }]
+    },
+    {
+        code: "(b) => b",
```
Comment (kaicataldo) [inline tests/lib/rules/arrow-parens.js]: This looks like a duplicate of the previous test - am I missing something?

File: tests/lib/rules/arrow-parens.js [hunk 15]
```diff
@@ -133,6 +156,44 @@ var invalid = [
...
+        output: "b => b",
+        options: ["as-needed", {requireForBlockBody: true}],
+        parserOptions: { ecmaVersion: 6 },
+        errors: [{
+            line: 1,
+            column: 1,
+            message: requireForBlockBodyMessage,
+            type: type
+        }]
     }
```
Comment (kaicataldo) [inline tests/lib/rules/arrow-parens.js]: Could we add a test that this is invalid: `(a) => ({});`. Don't mean to harp on that one case, but it seems like one that is more likely to break if someone ever touches this in the future. Would also love to see a test for destructured params in the invalid tests!

**Comment by kaicataldo** [comment]
Left some suggestions. Thanks so much for working on this @nfroidure!

**Comment by kaicataldo** [comment]
@nfroidure Just wanted to check in and see if there's anything I can do to help get this landed. Looks like we need to rebase and resolve a merge conflict as well.

**Comment by nfroidure** [comment]
@kaicataldo everything is well just had no time to dive back in this PR. I'll try to give it a look this week.

**Comment by nfroidure** [comment]
@kaicataldo just adapted the PR to your recommendations and rebased:
- about the duplicate it already was on the previous tests and was not sure of its usefulness. I removed both.
- i added a test for the `() => ({})` case.
- was not aware of the BlockStatement node type and its implications, just replaced it and it actually fixed the edge case you pointed out, nice catch ;).
- both {a, b} => a + b and [a, b] => a + b leads to a syntax error so i did not test them. Just added some tests to check ({a, b}) => a + b and ([a, b]) => a + b as valid.

**Comment by ljharb** [comment]
@nfroidure did you force push it up? the branch check still indicates conflicts

**Comment by nfroidure** [comment]
@ljharb i pushed it before rebasing to give you the opportunity to review the changes before the rebase but it appeared to be a minor conflict ;)

File: lib/rules/arrow-parens.js [hunk 16]
```diff
@@ -29,9 +38,13 @@ module.exports = {
         let message = "Expected parentheses around arrow function argument.";
         let asNeededMessage = "Unexpected parentheses around single function argument.";
         let asNeeded = context.options[0] === "as-needed";
+        let requireForBlockBodyMessage = "Unexpected parentheses around single function argument having a body with no curly braces";
+        let requireForBlockBodyNoParensMessage = "Expected parentheses around arrow function argument having a body with curly braces.";
+        let requireForBlockBody = asNeeded && context.options[1] && context.options[1].requireForBlockBody === true;
 
         let sourceCode = context.getSourceCode();
 
+
```
Comment (kaicataldo) [inline lib/rules/arrow-parens.js]: Extra newline

**Comment by kaicataldo** [comment]
Looks like a rebase is needed, but otherwise this is looking great to me.

**Comment by nzakas** [comment]
@kaicataldo feel free to merge if you think this is ready.

**Comment by kaicataldo** [comment]
LGTM. Thanks for contributing, @nfroidure!

**Comment by nfroidure** [comment]
Was a great pleasure to help on a product i use on a daily basis. Keep up the good work guys 😘
